### PR TITLE
Graduate DefaultPodTopologySpread to beta

### LINKED ISSUE #94863: Graduate DefaultPodTopologySpread to Beta
kubernetes/enhancements#1258

/sig scheduling

**Why is this needed**:

It allows to replace SelectorSpread plugin with a configurable default for the PodTopologySpread plugin.

This is important not just to reduce maintenance burden, but also because the new plugin provides better spreading, both at node level and bigger topologies. Since the score considers all the pods in the cluster, it is able to maintain zone balance even at 5k-node clusters.

Performance benchmarks will be used to ensure scalability is not hurt for vanilla pods.

### PR DESCRIPTION
**Pull request by alculquicondor**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Graduate DefaultPodTopologySpread to beta and enable the feature gate by default.

**Which issue(s) this PR fixes**:

Fixes #94863

kubernetes/enhancements#1258

**Special notes for your reviewer**:

Implementation requirements are fulfilled
https://git.k8s.io/enhancements/keps/sig-scheduling/1258-default-pod-topology-spread#beta-v120

Pending integration and conformance tests

**Does this PR introduce a user-facing change?**:

```release-note
DefaultPodTopologySpread graduated to Beta. The feature gate is enabled by default.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
- [KEP]: https://git.k8s.io/enhancements/keps/sig-scheduling/1258-default-pod-topology-spread
- [Usage]: https://kubernetes.io/docs/concepts/workloads/pods/pod-topology-spread-constraints/#cluster-level-default-constraints
```

### CODE REVIEW

**Comment by Huang-Wei** [comment]
LGTM. Please fix the CI failures.

**Comment by Huang-Wei** [comment]
Thanks @alculquicondor .

/lgtm

File: pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go [hunk 1]
```diff
@@ -664,10 +664,12 @@ func TestDryRunPreemption(t *testing.T) {
 				st.RegisterBindPlugin(defaultbinder.Name, defaultbinder.New),
 			)
 			registeredPlugins = append(registeredPlugins, tt.registerPlugins...)
+			informerFactory := informers.NewSharedInformerFactory(clientsetfake.NewSimpleClientset(), 0)
 			fwk, err := st.NewFramework(
 				registeredPlugins,
 				frameworkruntime.WithPodNominator(internalqueue.NewPodNominator()),
 				frameworkruntime.WithSnapshotSharedLister(snapshot),
+				frameworkruntime.WithInformerFactory(informerFactory),
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go:672]: why is this necessary now?

File: pkg/scheduler/apis/config/scheme/scheme_test.go [hunk 2]
```diff
@@ -302,7 +298,7 @@ profiles:
 						{
 							Name: "PodTopologySpread",
 							Args: &config.PodTopologySpreadArgs{
-								DefaultingType: config.ListDefaulting,
```
Comment (ahg-g) [inline pkg/scheduler/apis/config/scheme/scheme_test.go:305]: in retrospect, this looks strange, defaulting to listdefaulting without constraints.

**Review by ahg-g — COMMENTED** [review]
/lgtm

we should keep an eye on http://perf-dash.k8s.io/ to see how this will impact the scheduling latency.

**Comment by ahg-g** [comment]
/hold

just so you respond to the comments.

File: pkg/scheduler/apis/config/scheme/scheme_test.go (hunk 2 above)
Comment (ahg-g) [reply pkg/scheduler/apis/config/scheme/scheme_test.go:305]: oh, or is that how we disable the default constraints?

File: pkg/scheduler/apis/config/v1beta1/defaults_test.go [hunk 3]
```diff
@@ -390,38 +390,17 @@ func TestPluginArgsDefaults(t *testing.T) {
 						MaxSkew:           2,
 					},
 				},
+				// TODO(#94008): Make SystemDefaulting in v1beta2.
```
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/scheduler/apis/config/v1beta1/defaults_test.go:393]: Correct.

File: pkg/scheduler/apis/config/scheme/scheme_test.go (hunk 2 above)
Comment (alculquicondor) [reply pkg/scheduler/apis/config/scheme/scheme_test.go:305]: Correct. And that's the default if the feature gate is disabled.

File: pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go (hunk 1 above)
Comment (alculquicondor) [reply pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go:672]: Because when default spreading is enabled, we require the informers for Services and so on.

**Comment by alculquicondor** [comment]
I also need an approval for pkg/features

/hold cancel

File: pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go (hunk 1 above)
Comment (ahg-g) [reply pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go:672]: ok, and we were not using selectorspread in the tests (which would have required this).

Comment (alculquicondor) [reply pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go:672]: /hold
I didn't stop to think about that. Let me check

Comment (alculquicondor) [reply pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go:672]: /hold cancel

There is a test for preemption when there is hard topology spreading, which now has constraints by default.

Comment (Huang-Wei) [reply pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go:672]: Yes, the test that registers PodTopologySpread plugin now has a system default constraints, and it can pass only with a non-nil informerFactory::

https://github.com/kubernetes/kubernetes/blob/af1c8935d9f1f42767c6f5211a668e44a15946b4/pkg/scheduler/framework/plugins/podtopologyspread/plugin.go#L96-L99
